content/BCPatterns/generic-method-pattern/index.md

```diff
@@ -0,0 +1,327 @@
+    var
+        Handled: Boolean;
+    begin
+        ...
+        DoBlockCustomer(Cust, Handled);
+        ...
+    end;
+
+    local procedure DoBlockCustomer(var Cust: Record Customer; Handled: Boolean);
+    begin
+        if Handled then
+            exit;
+
+        Cust.Blocked := Cust.Blocked::All;
+        Cust.Modify(true);
+    end;
+...
+}
+```
+Usually indicated with a "do"-function, the business logic takes place in that procedure.  Obviously, when you have a decent amount of code, it's recommended that you make it readable by applying all the Best Practices in terms of readability in the codeunit.  Though, a few pointers here:
+- keep the cyclomatic complexity low
+  - one line (functioncall) after an IF-clause
+  - one line (functioncall) after a repeat
+- readable function calls
+- ...
+
+**Call out to the method from a table**
+
+Now, there is a reason that the one global procedure in the method-codeunit is `internal`.  We shouldn't be calling this procedure directly.  We should be calling it through it's "class" (usually, a table can be seen as a class).  That means, we would have to create a public (not internal) procedure on table-level, which can be used in the business logic.
+
+In our example above, this could be that table extension:
+
+```AL
+tableextension 53100 "Customer Ext BASE" extends Customer
+{
+    procedure BlockCustomer(HideDialog: Boolean)
+    var
+        WLDBlockCustomerMeth: Codeunit "WLD BlockCustomer Meth";
+    begin
+        WLDBlockCustomerMeth.BlockCustomer(Rec, HideDialog);
+    end;
+
+}
+```
+
+This practice improves readability.  In fact, by doing this, you just extended the suggesions-list in VSCode (IntelliSense) indicating a new method that your class can do.  This is very convenient for the developer that might need your new method.
+
+*Note - it could very well be that there simply isn't any table that can act as a class for our method.  In that case, you could use a codeunit as well.*
+
+**Naming Conventions**
+
+You might have noticed that the naming of our method is quite strict:
+- codeunit name: `WLD BlockCustomer Meth`
+- internal proc: `BlockCustomer`
+- do-procedure: `DoBlockCustomer`
+
+It is important to align these namings.  It indicates that the codeunit only does one thing (remember: encapsulation), and it improves searchability from outside the codeunit (for example when you're searching symbols or something).
+
+## Usage: 
+
+TODO
+
+## Benefits:
+
+As I said, it will facilitate a lot of advantages.  Let's explain a bit more in depth:
+
+### Extensibility
+Thanks to the *event layer*, by applying this pattern for all methods, we will automatically have the bare minimum of events that we need to hook into a method: the `OnBefore-` and the `OnAfter`.  Of course it would make sense to even add more events to the method when it would make sense (eg, when you're inserting a record in a table, it might be interesting to also raise an event just before you call the insert).
```

Comment:
"it would make sense" twice in the line. Suggestion to rephrase to:

Of course it makes sense to add more events to the method in certain situations. E.g. When you're inserting a record in a table, it might be a good idea to raise an event just before you call the insert.